### PR TITLE
Adding threshold_multiotsu.

### PR DESCRIPTION
**Pull request by alexandrejaguar**
## Description
Adding the function `threshold_multiotsu()`. This generates multiple thresholds for an input image, based on the Multi-Otsu implementation by Liao and Chung.

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [X] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [X] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [ ] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[1] Liao, P-S. and Chung, P-C., "A fast algorithm for multilevel thresholding", Journal of Information Science and Engineering 17 (5): 713-727, 2001.

### CODE REVIEW

**Comment by soupault** [comment]
@alexandrejaguar Thanks for the valuable contribution!
From my point of view it would be nice to add also some tests and, more important, gallery example with multiple thresholds.
Once you manage to do that, post a comment here and I'll take a thorough look.

**Comment by soupault** [comment]
@alexandrejaguar regarding your question on tests in gitter discussion:
First, please, take a look at the existing tests in https://github.com/scikit-image/scikit-image/blob/master/skimage/filters/tests/test_thresholding.py .
Then, I'd suggest to implement (a) couple of simple tests checking is function returns error on invalid input, (b) checks on corner cases (for example, with monotonic input image), (c) checks with input images we're already shipping (e.g. astronaut, coins - one is enough). This will do the basic testing pretty much.

Speaking of the gallery example, I think it would be better to target an example as catchy as possible (select the image with the most indicative results, etc).

Should you have any questions, feel free to continue discussion here.

File: doc/examples/segmentation/plot_multiotsu.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+"""
+=======================
+Multi-Otsu Thresholding
+=======================
+
+In this example we use multi-Otsu threshold to separate the pixels of an
```
Comment (sciunto) [inline doc/examples/segmentation/plot_multiotsu.py]: Multi-otsu threshold is used to separate [...] (This is for the online preview)

File: doc/examples/segmentation/plot_multiotsu.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+# Stablishing the font size for all plots.
+matplotlib.rcParams['font.size'] = 9
+
+# The input image.
+image = camera()
+
+# Applying multi-Otsu threshold for the default value, generating
+# three classes.
+thresh = threshold_multiotsu(image)
+print('Threshold values: ', thresh)
```
Comment (sciunto) [inline doc/examples/segmentation/plot_multiotsu.py]: I don't think we want a print here. However, you can display that on a histogram. You can use

```
ax.axvline(thresh[0], color='r')
```
for example

Comment (sciunto) [inline doc/examples/segmentation/plot_multiotsu.py]: I just saw you did it below. So, you can only remove the print

File: skimage/filters/thresholding.py [hunk 3]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
             new_thresh = temp + tolerance
 
     return threshold + immin
+
+
+def threshold_multiotsu(image, nclass=3, nbins=256):
+    """Generates multiple thresholds for an input image. Based on the
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: Break after the point. Only one sentence for the first line.

File: skimage/filters/thresholding.py [hunk 4]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
...
+    Multi-Otsu implementation by Liao and Chung.
+
+    Parameters
+    ----------
+    image : array
+        Grayscale input image.
+    nclass : int, optional
+        Number of classes to be thresholded, i.e. the number of resulting
+        regions. Accepts an integer from 2 to 5.
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: Can we check that nclass is in [2,5] in the code? And raise ValueError if not correct.

File: skimage/filters/thresholding.py [hunk 5]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
...
+        raise TypeError("Your image seems to be RGB (shape: {0}. Please use a"
+                        "grayscale image.".format(image.shape))
+
+    # check if the image has more than one color.
+    if image.min() == image.max():
+        raise TypeError("The input image seems to have just one color: {0}."
+                        "Please use a grayscale image.".format(image.min()))
+
+    # image needs to be treated as float.
+    image = img_as_float(image)
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: import problem here (that's why travis failed)

File: skimage/filters/thresholding.py [hunk 6]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
...
+
+    # calculating the histogram and the probability of each gray level.
+    hist, _ = histogram(image.ravel(), nbins)
+    prob = hist/np.size(image)
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: spaces around /

File: skimage/filters/thresholding.py [hunk 7]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
...
+    for u in range(2, gray_lvl):
+        for v in range(u, gray_lvl):
+            momP[u, v] = momP[1, v] - momP[1, u-1]
+            momS[u, v] = momS[1, v] - momS[1, u-1]
+
+    # building the lookup tables: calculating the between class variance.
+    for u in range(1, gray_lvl):
+        for v in range(u+1, gray_lvl):
+            if (momP[u, v] != 0):
+                var_btwcls[u, v] = (momS[u, v]**2)/momP[u, v]
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: spaces around /

File: skimage/filters/thresholding.py [hunk 8]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
...
+            else:
+                var_btwcls[u, v] = 0
+
+    # finding max threshold candidates, depending on nclass.
+    # number of thresholds is equal to number of classes - 1.
+    if nclass == 2:
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: Correct me if I'm wrong, but this case should match htreshold_otsu(). This could be a test.

File: skimage/filters/thresholding.py [hunk 9]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
...
     return threshold + immin
+
+
+def threshold_multiotsu(image, nclass=3, nbins=256):
+    """Generates multiple thresholds for an input image. Based on the
+    Multi-Otsu implementation by Liao and Chung.
+
+    Parameters
+    ----------
+    image : array
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: @soupault would love to have dimensions here :)

File: skimage/filters/thresholding.py [hunk 10]
```diff
@@ -384,3 +385,137 @@ def threshold_li(image):
...
+                        "Please use a grayscale image.".format(image.min()))
+
+    # image needs to be treated as float.
+    image = img_as_float(image)
+
+    # calculating the histogram and the probability of each gray level.
+    hist, _ = histogram(image.ravel(), nbins)
+    prob = hist/np.size(image)
+
+    gray_lvl = 256
```
Comment (sciunto) [inline skimage/filters/thresholding.py]: Here, we make an assumption on the dtype of the input image, no?

**Comment by sciunto** [comment]
THat's a very cool feature. I made few comments.

The PR needs a rebase and some tests. Then, we should be good. :)

**Comment by alexandrejaguar** [comment]
Hi @sciunto, thank you very much for your comments!
I am going on a big trip these days, but I'll correct that as soon as I get back home.
Thanks again!

**Comment by sciunto** [comment]
np. Take care!

File: skimage/filters/thresholding.py (hunk 8 above)
Comment (alexandrejaguar) [inline skimage/filters/thresholding.py]: @sciunto I made a test today, and this does not happen! I'll check the code and correct it as soon as possible.

**Comment by sciunto** [comment]
You also need to list your function in skimage/filters/__init__.py

File: skimage/filters/thresholding.py (hunk 10 above)
Comment (alexandrejaguar) [inline skimage/filters/thresholding.py]: @sciunto I could not solve this problem. I tried to reproduce the behavior of Tosa's plugin: http://imagej.net/plugins/download/Multi_OtsuThreshold.java
However, I didn't figure out yet heow to divide the classes according to the kind of the input image... for now, this code is useless.

**Comment by alexandrejaguar** [comment]
Hi all,
I had some improvements, but this function is not reliable yet.
I contacted Tosa today. I am waiting his response (and maybe some help).
Actually, if you could help, there is the newest version of the function. Please take a look:
https://github.com/alexandrejaguar/software/blob/master/threshold_multiotsu.py
Thanks!

**Comment by sciunto** [comment]
Hi @alexandrejaguar 

What do you mean by not reliable? Perhaps you can rebase and update this branch and shows what's wrong with some tests.

**Comment by sciunto** [comment]
Hi @alexandrejaguar 

What do you mean by not reliable? Perhaps you can rebase and update this branch and shows what's wrong with some tests.

**Comment by alexandrejaguar** [comment]
Hi @sciunto,
maybe there's something wrong, but I can't commit code to `threshold_multiotsu`.
I'm trying to solve that. Thanks!

**Comment by sciunto** [comment]
@alexandrejaguar What's your commands, errors and .git/config content?

**Comment by alexandrejaguar** [comment]
Hi @sciunto,

* commands:
```
git checkout -b threshold_multiotsu
git add skimage/filters/thresholding.py
git add doc/examples/segmentation/plot_multiotsu.py
git commit -m 'Adding @sciunto corrections and improving code.'
git push origin threshold_multiotsu
```

* issues: commands return everything is ok, but I can't see the contribution.

** on `commit`:
```
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean
```

** on `push`:

```
Username for 'https://github.com': alexandrejaguar
Password for 'https://alexandrejaguar@github.com': 
Everything up-to-date
```

* `.git/config:`
```
[core]
        repositoryformatversion = 0
        filemode = true
        bare = false
        logallrefupdates = true
[remote "origin"]
        url = https://github.com/alexandrejaguar/scikit-image.git
        fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
[remote "upstream"]
        url = https://github.com/scikit-image/scikit-image.git
        fetch = +refs/heads/*:refs/remotes/upstream/*
```

Thank you very much for your help!

**Comment by sciunto** [comment]
I think you did a mistake. You pushed on alexandrejaguar/master instead of alexandrejaguar/threshold_multiotsu (You probably did git origin push by mistake).

I can see your commit here: https://github.com/alexandrejaguar/scikit-image
but not there: https://github.com/alexandrejaguar/scikit-image/tree/threshold_multiotsu

You need to clean up your master branch. I think https://stackoverflow.com/questions/9646167/clean-up-a-fork-and-restart-it-from-the-upstream or https://stackoverflow.com/questions/8134960/how-to-revert-master-branch-to-upstream is what you are looking for.

Your command lines above looks good to me.

**Comment by alexandrejaguar** [comment]
Thank you very much. I resetted my repo using https://stackoverflow.com/questions/9646167/clean-up-a-fork-and-restart-it-from-the-upstream; let's see if I can redo it now.
Thanks again!

**Comment by sciunto** [comment]
Your PR also needs a rebase a little bit tricky. So, if you prefer to start a new PR because it's easier, I'm fine with it ;)

**Comment by alexandrejaguar** [comment]
@sciunto hahaha! The time has passed, and i need to learn how to do that things.
How painful would each alternative be? Could you help me on what I need to do?
Thanks for all your help!

**Comment by sciunto** [comment]
It's easy to make a new PR. Your code is really entangled with the recent evolution of master. I looked, it's doable, but probably to difficult for the first time ;)

If you want to create a new PR. sync master, create a new branch, a copy paste your code and do as usual.

**Comment by sciunto** [comment]
@alexandrejaguar How do you want to proceed?
